### PR TITLE
Do not decompose domain along user-specified dimensions

### PR DESCRIPTION
**Pull request by spossann**
Add new flag `mpi_dims_mask` to `discretize_domain`, which is successively passed to:
- `ddm.cart.DomainDecomposition`
- `ddm.partition.compute_dims`
- `ddm.partition.compute_dims_general`

Function `compute_dims_general` is the ones responsible to create the Cartesian grid of MPI processes. It receives `mpi_dims_mask` which is either `None` or a list of `bool`. In the latter case each element is set to `True` if the dimension is to be used in the domain decomposition (=default for each dimension). If `mpi_dim_mask[i] == False`, the i-th dimension is not decomposed.

Two unit tests for 2d and 3d have been added.

### CODE REVIEW

File: psydac/ddm/partition.py [hunk 1]
```diff
@@ -132,10 +136,21 @@ def compute_dims( nnodes, gridsizes, min_blocksizes=None, mpi=None, try_uniform=
     return dims, blocksizes
 
 #==============================================================================
-def compute_dims_general( mpi_size, npts ):
+def compute_dims_general( mpi_size, npts, dims_mask=None ):
+
+    if dims_mask is None:
+        dims_mask = [True]*len( npts )
+    else:
+        assert any(dims_mask), 'dims_mask must contain at least one True value.'
```
Comment (yguclu) [inline psydac/ddm/partition.py:144]: I suggest a couple more checks here:
```suggestion
    if dims_mask is None:
        dims_mask = [True] * len(npts)
    else:
        assert len(dims_mask) == len(npts), "dims_mask must have one entry for each dimension."
        assert all(isinstance(m, bool) for m in dims_mask), "dims_mask must only contain True/False values."
        assert any(dims_mask), "dims_mask must contain at least one True value."
```

File: psydac/ddm/tests/test_partition.py [hunk 2]
```diff
@@ -48,6 +48,30 @@ def test_partition_1d_general( mpi_size ):
...
+def test_partition_2d_dims_mask( mpi_size, npts, mask ):
+
+    print(f'\nmpi_size={mpi_size}:')    
+    # General partition: blocks are not all identical but closer to a cube
+    dims, blocksizes = compute_dims( mpi_size, npts, [3,3] )   
+    print(f'general:'.ljust(30), dims, blocksizes) 
+    
+    # Mask dimensions
+    dims, blocksizes = compute_dims( mpi_size, npts, [3,3], dims_mask=mask )   
+    print(f'dims_mask {mask}:'.ljust(30), dims, blocksizes)
```
Comment (yguclu) [inline psydac/ddm/tests/test_partition.py:65]: Although we still have print statements in some of our tests, Pytest is not happy about it. Here you should pass the flag `verbose = False` to the new functions, and only print to terminal if this flag is set to `True`. Then you can pass `verbose = True` when you run the script manually (bottom of the file).

Or you can simply remove the print statements, of course.

**Review by yguclu — CHANGES_REQUESTED** [review]
Thanks @spossann, this looks good to me!

Although I wonder if we could find a better name for the `dims_mask` parameter in function `discretize_domain`. The current name may suggest the user that such a dimension is not discretized, while this is not the case. Maybe `mpi_dims_mask` is more telling?

@saidctb What do you think?

File: psydac/ddm/partition.py (hunk 1 above)
Comment (spossann) [reply psydac/ddm/partition.py:144]: Agreed. Should I do "commit suggestion"?

File: psydac/ddm/tests/test_partition.py (hunk 2 above)
Comment (spossann) [reply psydac/ddm/tests/test_partition.py:65]: Ok I will add `verbose=False`. What do you mean by Pytest does not like it? Can there be problems?

Comment (yguclu) [reply psydac/ddm/tests/test_partition.py:65]: By default Pytest collects the standard output and only shows it if there is an error or warning. This means that if a test fails the error message will be cluttered with the output from all the other tests (which do not fail). This is far from ideal.

File: psydac/ddm/partition.py (hunk 1 above)
Comment (yguclu) [reply psydac/ddm/partition.py:144]: If you like my suggestion, that is the fastest way to include it.

**Comment by yguclu** [comment]
The workflow with Python 3.7 fails because of a download problem, I am running it again.

Codacy raises an error that does not make sense to me...

**Review by yguclu — APPROVED** [review]

**Comment by yguclu** [comment]
> Can we rerun the Codacy test? The error came with the commit [4c4385f](https://github.com/pyccel/psydac/commit/4c4385fe07b4944373a1e18a871fc8390b0ae491), which was however resolved by the later commits.

I have rerun them 10 times already, but Codacy only processes the files that have been changed in the last commit. So it keeps raising the same error. I deactivated the Codacy test temporarily, so we can merge this.

**Pull request merged**